datastream-kafka/src/test/java/com/linkedin/datastream/kafka/TestKafkaProducerWrapper.java

```diff
@@ -0,0 +1,147 @@
+
+    DatastreamTask task = new DatastreamTaskImpl(Collections.singletonList(ds));
+    ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(topicName, null, null);
+    producerWrapper.assignTask(task);
+
+    // Sending first event, send should pass, none of the other methods on the producer should have been called
+    producerWrapper.send(task, producerRecord, null);
+    producerWrapper.verifySend(1);
+    producerWrapper.verifyFlush(0);
+    producerWrapper.verifyClose(0);
+    Assert.assertEquals(producerWrapper.getNumCreateKafkaProducerCalls(), 1);
+
+    ExecutorService executorService = Executors.newCachedThreadPool();
+    executorService.submit(() -> {
+      // Flush has been mocked to throw an InterruptException
+      Assert.assertThrows(InterruptException.class, producerWrapper::flush);
+    }).get();
+
+    producerWrapper.verifySend(1);
+    producerWrapper.verifyFlush(1);
+    producerWrapper.verifyClose(1);
+
+    // Second send should create a new producer, resetting flush() and close() invocations count
+    producerWrapper.send(task, producerRecord, null);
+    producerWrapper.verifySend(1);
+    producerWrapper.verifyFlush(0);
+    producerWrapper.verifyClose(0);
+    Assert.assertEquals(producerWrapper.getNumCreateKafkaProducerCalls(), 2);
+
+    // Second producer's flush() has not been mocked to throw exceptions, this should not throw
+    producerWrapper.flush();
+    producerWrapper.verifySend(1);
+    producerWrapper.verifyFlush(1);
+    producerWrapper.verifyClose(0);
+    Assert.assertEquals(producerWrapper.getNumCreateKafkaProducerCalls(), 2);
+
+    // Send should reuse the older producer and the counts should not be reset
+    producerWrapper.send(task, producerRecord, null);
+    producerWrapper.verifySend(2);
+    producerWrapper.verifyFlush(1);
+    producerWrapper.verifyClose(0);
+    Assert.assertEquals(producerWrapper.numCreateKafkaProducerCalls, 2);
+
+    // Closing the producer's task, and since this is the only task, the producer should be closed
+    producerWrapper.close(task);
+    producerWrapper.verifySend(2);
+    producerWrapper.verifyFlush(1);
+    producerWrapper.verifyClose(1);
+    Assert.assertEquals(producerWrapper.numCreateKafkaProducerCalls, 2);
+  }
+
+  private static class MockKafkaProducerWrapper extends KafkaProducerWrapper<byte[], byte[]> {
+    private boolean createKafkaProducerCalled;
+    private int numCreateKafkaProducerCalls;
+    private Producer<byte[], byte[]> mockProducer;
+
+    MockKafkaProducerWrapper(String logSuffix, Properties props, String metricsNamesPrefix) {
+      super(logSuffix, props, metricsNamesPrefix);
+    }
+
+    @Override
+    Producer<byte[], byte[]> createKafkaProducer() {
+      @SuppressWarnings("unchecked")
+      Producer<byte[], byte[]> producer = (Producer<byte[], byte[]>) mock(Producer.class);
+      if (!createKafkaProducerCalled) {
```

Comment:
Just to make it super simple for anyone reading this test, please add a comment summarizing your intent (e.g. `calling flush() on the first created producer will throw`).

Reply:
done